### PR TITLE
Don't apply swap cast optimization when explicit cast of the column is used

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Resolves https://github.com/crate/crate/issues/12135

Replaces https://github.com/crate/crate/pull/12153

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by BaurzhanSakhariev** [comment]
infrastructure related failure

retest this please

File: server/src/test/java/io/crate/planner/optimizer/symbol/OptimizerTest.java [hunk 1]
```diff
@@ -43,4 +45,24 @@ public void test_like_on_numeric_columns_keeps_cast_around_reference() throws Ex
...
+
+    @Test
+    public void test_comparison_of_string_and_numeric_doesnt_optimize_away_cast() throws Exception {
+        SQLExecutor e = SQLExecutor.builder(clusterService)
+            .addTable("create table tbl (strCol string, intCol int)")
+            .build();
+
+        //checking either sides of num <-> str comparison, reference casting remains
+        Symbol symbol1 = Optimizer.optimizeCasts(e.asSymbol("strCol::bigint > 3"), e.getPlannerContext(clusterService.state()));
+        Symbol symbol2 = Optimizer.optimizeCasts(e.asSymbol("intCol::string > '3'"), e.getPlannerContext(clusterService.state()));
```
Comment (seut) [inline server/src/test/java/io/crate/planner/optimizer/symbol/OptimizerTest.java:57]: is there a difference for `::string` and `::bigint` casts or why do we have to test both?

File: server/src/test/java/io/crate/planner/optimizer/symbol/OptimizerTest.java [hunk 2]
```diff
@@ -43,4 +45,24 @@ public void test_like_on_numeric_columns_keeps_cast_around_reference() throws Ex
...
+
+        assertThat(symbol1, SymbolMatchers.isFunction(GtOperator.NAME));
+        assertThat(symbol2, SymbolMatchers.isFunction(GtOperator.NAME));
+
+        Function func1 = (Function) symbol1;
+        Function func2 = (Function) symbol1;
+        // In all cases below ref cast used to be swapped with parameter.
+        assertThat(func1.arguments().get(0), SymbolMatchers.isFunction("cast"));
+        assertThat(func2.arguments().get(0), SymbolMatchers.isFunction("cast"));
```
Comment (seut) [inline server/src/test/java/io/crate/planner/optimizer/symbol/OptimizerTest.java:66]: minor: wouldn't something like that be a bit simpler?
```
assertThat(symbol2, SymbolMatchers.isFunction(
            GtOperator.NAME,
            SymbolMatchers.isFunction("cast"),
            SymbolMatchers.isLiteral("3")
        ));
```

File: docs/appendices/release-notes/unreleased.rst [hunk 3]
```diff
@@ -82,4 +82,10 @@ Fixes
   is not used in an upper query. For example:
   ``select x from (select x, ROW_NUMBER() OVER (PARTITION BY y) from t) t1``
 
+- Fixed an issue that caused numbers being compared as strings in queries where
+  selected column is string casted to number and compared with a number or vice
+  versa. Example:
+  ``SELECT * FROM test WHERE strCol::bigint > 3`` or
+  ``SELECT * FROM test WHERE intCol::string > '3'``
```
Comment (seut) [inline docs/appendices/release-notes/unreleased.rst:89]:
```suggestion
- Fixed an incorrect optimization of comparison function arguments for explicit
  cast arguments resulting in a wrong result set. Example:
   ```WHERE strCol::bigint > 3```
```

File: server/src/test/java/io/crate/planner/optimizer/symbol/OptimizerTest.java (hunk 1 above)
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/planner/optimizer/symbol/OptimizerTest.java:57]: no difference, artefact from the previous implementation where I played around with (num,str)/(str,num) and wanted to test explicitly. will remove second test, for this implementation is redundant.

File: server/src/test/java/io/crate/planner/optimizer/symbol/OptimizerTest.java (hunk 2 above)
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/planner/optimizer/symbol/OptimizerTest.java:66]: yep, changed that, thanks. 

I wanted to check that precisely `arg(0)/ ref` is being casted, had an impression, that `isFunction` just calls contains which could lead to false test success in `strCol::bigint > 3::bigint ` even with old code without fix. I reverted fix locally and test this assumption, I was wrong, so committing simpler assertion

**Review by seut — APPROVED** [review]
👍 good catch

**Comment by BaurzhanSakhariev** [comment]
`crate.client.exceptions.ConnectionError: No more Servers available, exception from last server`

not related, retest this please

**Pull request merged**